proxy/server/server.go

```diff
@@ -357,3 +378,133 @@ func newScopeRouter(scope string, router http.Handler) (http.Handler, error) {
 	}
 	return scopedRouter, nil
 }
+
+// Returns a certificate based on a hosts KeyPair settings.
+func certForHost(hostCfg engine.Host) (tls.Certificate, error) {
+	c := hostCfg.Settings.KeyPair
+	cert, err := tls.X509KeyPair(c.Cert, c.Key)
+	if err != nil {
+		return tls.Certificate{}, err
+	}
+	return cert, nil
+}
+
+// Returns a GetCertificate function for this host based on AutoCert settings,
+// by generating an autocert manager for the host.
+// It optionally will staple the OCSP response to the cert when it is generated,
+// if OCSP stapling is enabled for the host.
+func certFuncForHost(hostCfg engine.Host, autoCertCache autocert.Cache, stapler stapler.Stapler) (getCertificateFunc, error) {
+	ac := hostCfg.Settings.AutoCert
+
+	// Each host gets its own Autocert Manager - this allows individual
+	// certs to use different autocert authorities, as well as auth keys
+	autoCertMgr := &autocert.Manager{
+		Prompt:     autocert.AcceptTOS,
+		Cache:      autoCertCache,
+		HostPolicy: autocert.HostWhitelist(hostCfg.Name),
+		Email:      ac.Email,
+	}
+
+	if ac.RenewBeforeSeconds > 0 {
+		autoCertMgr.RenewBefore = time.Duration(ac.RenewBeforeSeconds) * time.Second
+	}
+
+	// if either directory or key are non-empty, we need to generate
+	// a custom ACME client to override either.
+	if ac.Key != "" || ac.DirectoryURL != "" {
+
+		// If DirectoryURL is empty, the default Let's Encrypt URL will be picked.
+		autoCertMgr.Client = &acme.Client{
+			DirectoryURL: ac.DirectoryURL,
+		}
+
+		// If Key is non-empty, then decode it as RSA or EC which are the only two keys
+		// we support. Go's crypto library doesn't support a generic function to provide back
+		// a private key interface.
+		if ac.Key != "" {
+			block, _ := pem.Decode([]byte(ac.Key))
+			if block == nil {
+				return nil, fmt.Errorf("Autocert Key PEM Block for Host %s is invalid.", hostCfg.Name)
+			} else if block.Type == "RSA PRIVATE KEY" {
+				rsaPrivateKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+				if err != nil {
+					autoCertMgr.Client.Key = rsaPrivateKey
```

Comment:
if error then assign?

Reply:
Added two UTs to guard against both code paths - RSA and EC Private Keys.